www_som/res_partner.py

```diff
@@ -275,6 +275,69 @@ def _www_soci(self, cursor, uid, ids, field_name, arg, context=None):
                     res[partner_id] = empty_num_soci
         return res
 
+    def www_ov_attachments(self, cursor, uid, partner_id, context=None):
+        """
+        Returns the list of the OV attachments for this parnter.
+        TODO: Now only works for attachments in the res_partner and poweremail.mailbox,
+        but it can be improved in the future.
+        """
+        attachment_obj = self.pool.get("ir.attachment")
+        attachment_ids = self._query_ov_attachments_ids(cursor, uid, partner_id)
+        res = []
+        for attachment in attachment_obj.browse(cursor, uid, attachment_ids):
+            res.append({
+                'erp_id': attachment.id,
+                'name': attachment.name,
+                'file_name': attachment.datas_fname,
+                'date': attachment.create_date,
+                'res_model': attachment.res_model,
+                'res_id': attachment.res_id,
+                'category': attachment.category_id.code,
+            })
+        return res
+
+    def www_get_ov_attachment(self, cursor, uid, attachment_id, partner_id, context=None):
+        """
+        Downloads an OV attachment
+        TODO: Now only works for attachments in the res_partner and poweremail.mailbox,
+        but it can be improved in the future.
+        """
+        attachment_obj = self.pool.get("ir.attachment")
+        attachment_ids = self._query_ov_attachments_ids(
+            cursor, uid, partner_id, attachment_id)
+        if not attachment_ids:
+            raise Exception("The attachment doesn't exists or you don't have permission")
+
+        return attachment_obj.read(cursor, uid, attachment_id, ['datas_fname', 'datas'])
+
+
+    def _query_ov_attachments_ids(
+            self, cursor, uid, partner_id, attachment_id=None, context=None):
+        """
+        Gets the IDs of a partner OV attachments, also check a single attachment.
+        TODO: Now only works for attachments in the res_partner and poweremail.mailbox,
+        but it can be improved in the future.
+        """
+        cursor.execute('''
```

Comment:
@anxodio sql és menys mantenible. 
Estava intentant passar la query a ORM però em trobo que quan la faig retorna valors inesperats (els meus adjunts i tots els adjunts de certificats 4.000 i pico), a stp1.

```sql
select ia.* 
            from ir_attachment ia
            inner join ir_attachment_category iac on ia.category_id = iac.id
            where iac.ov_available is true
            and null is null or ia.id = null
            and (
             (ia.res_model = 'res.partner' and ia.res_id = 217)
            or
             (ia.res_model = 'poweremail.mailbox' and (
              select partner_id from som_enviament_massiu sem where sem.id = (
               select SPLIT_PART(pm.reference, ',', 2)::int from poweremail_mailbox pm
               where pm.id = ia.res_id
               and SPLIT_PART(pm.reference, ',', 1) = 'som.enviament.massiu'
             )) =217)
            );
```

Potser amb un test sortim de dubtes

Reply:
`# Params
attachment_id = None
partner_id = 217

# Query
models = ['som.enviament.massiu', 'poweremail.mailbox']
categories = O.IrAttachmentCategory.search([('ov_available','=',True)])
adjunts_de_partners = O.IrAttachment.search([('res_model','=','res.partner'),('res_id','=',part
ner_id), ('category_id','in', categories)])
`

La part dels adjunts de mails, la trobo perillosa, em sembla un model dèbil (fàcilment eliminable). Potser s'haurien de posar aquests adjunts a la fitxa del partner també.

Reply:
Parlat per meet.